internal/locate/replica_selector.go

```diff
@@ -0,0 +1,554 @@
+	maxScoreIdxes := make([]int, 0, len(replicas))
+	maxScore := -1
+	reloadRegion := false
+	for i, r := range replicas {
+		epochStale := r.isEpochStale()
+		liveness := r.store.getLivenessState()
+		isLeader := AccessIndex(i) == s.leaderIdx
+		if epochStale && ((liveness == reachable && r.store.getResolveState() == resolved) || isLeader) {
+			reloadRegion = true
+		}
+		if !s.isCandidate(r, isLeader, epochStale, liveness) {
+			continue
+		}
+		score := s.calculateScore(r, isLeader)
+		if score > maxScore {
+			maxScore = score
+			maxScoreIdxes = append(maxScoreIdxes[:0], i)
+		} else if score == maxScore && score > -1 {
+			maxScoreIdxes = append(maxScoreIdxes, i)
+		}
+	}
+	if reloadRegion {
+		selector.region.setSyncFlags(needDelayedReloadPending)
+	}
+	if len(maxScoreIdxes) == 1 {
+		idx := maxScoreIdxes[0]
+		return replicas[idx]
+	} else if len(maxScoreIdxes) > 1 {
+		// if there are more than one replica with the same max score, we will randomly select one
+		// todo: consider use store slow score to select a faster one.
+		idx := maxScoreIdxes[randIntn(len(maxScoreIdxes))]
+		return replicas[idx]
+	}
+	if s.busyThreshold > 0 {
+		// when can't find an idle replica, no need to invalidate region.
+		return nil
+	}
+
+	metrics.TiKVReplicaSelectorFailureCounter.WithLabelValues("exhausted").Inc()
+	for _, r := range replicas {
+		if r.deadlineErrUsingConfTimeout {
+			// when meet deadline exceeded error, do fast retry without invalidate region cache.
+			return nil
+		}
+	}
+	selector.invalidateRegion() // is exhausted, Is need to invalidate the region?
+	return nil
+}
+
+func (s ReplicaSelectMixedStrategy) isCandidate(r *replica, isLeader bool, epochStale bool, liveness livenessState) bool {
+	if epochStale || liveness == unreachable {
+		// the replica is not available, skip it.
+		return false
+	}
+	maxAttempt := 1
+	if r.dataIsNotReady && !isLeader {
+		// If the replica is failed by data not ready with stale read, we can retry it with replica-read.
+		// 	after https://github.com/tikv/tikv/pull/15726, the leader will not return DataIsNotReady error,
+		//	then no need to retry leader again, if you try it again, you may got a NotLeader error.
+		maxAttempt = 2
+	}
+	if r.isExhausted(maxAttempt, 0) {
+		// attempts is exhausted, skip it.
+		return false
+	}
+	if s.leaderOnly && !isLeader {
+		return false
+	}
+	if s.busyThreshold > 0 && (r.store.EstimatedWaitTime() > s.busyThreshold || r.serverIsBusy || isLeader) {
+		return false
+	}
+	return true
+}
+
+const (
+	// define the score of priority.
+	scoreOfLabelMatch   = 3
+	scoreOfPreferLeader = 2
+	scoreOfNormalPeer   = 1
+	scoreOfNotSlow      = 1
+)
+
+func (s *ReplicaSelectMixedStrategy) calculateScore(r *replica, isLeader bool) int {
+	score := 0
+	if isLeader {
+		if s.preferLeader {
+			score += scoreOfPreferLeader
+		} else if s.tryLeader {
+			if len(s.labels) > 0 {
+				// when has match labels, prefer leader than not-matched peers.
+				score += scoreOfPreferLeader
```

Comment:
* Should a not-slow normal follower (`NotSlow` + `NormalPeer` -> 2) have the same priority as a slow leader in prefer-leader mode (`PreferLeader` -> 2) ?

* Is there an assumption that in cases that leader must be expected (such as the first attempt in prefer-leader mode and the second attempt in stale read mode) must be done with the `ReplicaSelectLeaderStrategy` instead of using the `calculateScore`? If so, I think there should be some comments to explain that this function doesn't handle those cases.

* If `LabelMatch` always has higher priority than `PreferLeader`, and `PreferLeader` always has higher priority than `NormalPeer` and `NotSlow`, is it possible that we use different bits for these factors, so that the more significant bit a factor is at, the more dominant status the factor is? For factors that has the same priority, we reserve more bits for them. For example, we can have such definition:

    ```go
    const (
    	flagLabelMatches = 1 << 3
    	flagPreferLeader = 1 << 2
    	flagNormalPeer = 1
    	flagNotSlow = 1
    	
    	// So that we have definition of the final score:
    	// MSB                                                                               LSB
    	// [unused bits][1 bit: LabelMatches][1 bit: PreferLeader][2 bits: NormalPeer + NotSlow]
    )
    ```

    I'm not sure if this is possible to give the same result as your current code, but I think in this way, we can have less cases like:

   * Factor A < Factor C
   * Factor B < Factor C
   * Factor A + Factor B > Factor C

    which I'm afraid may make the priority calculation less comprehensible. Another benefit is that we can mostly know how a score comes from by reading its bits, which may help in diagnosing.

Reply:
Great catch, done as your advice.